### ISSUE TITLE
[Learning curve]Save training/test metric during training?

### ISSUE THREAD

**Issue by wangg12** [dmlc/mxnet#798]
I wonder how I can save those metrics during trainging and then I can plot the learning curve. Currently, the training process has its log but it's not so easy to get those values especially when the training epochs are very large. 

PS: I want to do it in pyhton.

**Comment by antinucleon** [comment]
You can use batch callback to get it out, eg
https://github.com/dmlc/mxnet/blob/master/python/mxnet/callback.py#L30
https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L199

**Comment by qiaohaijun** [comment]
I occur the same problem

**Comment by wangg12** [comment]
@antinucleon In that callback, there is just a log. How to get the train and val metric values? And another question is can we assign more than one callback to batch_end_callback or epoch_end_callback?

**Comment by antinucleon** [comment]
You can parse the log to get what you want, which is the way TensorBoard
does in TensorFlow.

**Comment by tqchen** [comment]
@wangg12 if you take a look at these callbacks, you can find it is not hard to create another closure to pass and push data into a list, for example,
```python
def log_to_list(period, lst):
    def _callback(param):
        """The checkpoint function."""
        if param.nbatch % period == 0:
            name, value = param.eval_metric.get()
            lst.append(value)
    return _callback
```
I think multiple callbacks can be passed in via a list

**Comment by wangg12** [comment]
@tqchen Thanks for your help! This works for train data metric, but how to apply to val data?

**Comment by wangg12** [comment]
And it seems we can only get batch end metrics in this way. I'm afraid parsing the log is the only way to get the epoch end metrics by now.

**Comment by paipai880429** [comment]
@tqchen Tianqi, thanks a lot for showing us such s simple example to clarify the use of callbacks of mxnet. Now I know how to get the training accuracy list so that I can plot in ```matplotlib```. One more question: where can I find the attributes of ```param```? What if I would like to log the validation accuracy and loss into another list? Besides, the document of mxnet needs improvement ^_^.

**Comment by paipai880429** [comment]
@tqchen @antinucleon @winstywang Looking forward to your answer. Thank you for your time!

**Comment by tqchen** [comment]
currently it was not possible mainly because the validation accuracy eval metric is not presented to the epoch end callback. We can make some changes to the training function to expose more environement into the callback.

**Comment by paipai880429** [comment]
@tqchen Thanks for your answer. Therefore, the best way to choose the model is the checkpoint the training log and analyze it by eyeball checking, right?
